src/main/java/net/starlark/java/eval/StarlarkInt.java

```diff
@@ -414,6 +439,25 @@ public static StarlarkInt add(StarlarkInt x, StarlarkInt y) {
       return StarlarkInt.of(xl + yl);
     }
 
+    try {
+      long xl = x.toLongFast();
+      long yl = y.toLongFast();
+      long r = xl + yl;
+      // overflow check
+      // x >= 0 | y >= 0 | r >= 0 | overflow?
+      // t      | t      | t      | f
+      // t      | t      | f      | t  <--
+      // f      | f      | t      | t  <--
+      // f      | f      | f      | f
+      // t      | f      | *      | f
+      // f      | t      | *      | f
+      boolean overflow = ((xl >= 0) != (r >= 0)) && ((yl >= 0) != (r >= 0));
```

Comment:
Use this shorter version from Hacker's Delight, which does the parity computations before the signed comparison:
```
     boolean overflow = ((xl ^ zl) & (yl ^ zl)) < 0; // see Hacker's Delight, chapter 2
```